File: src/main/java/de/n2online/sonification/Agent.java
package de.n2online.sonification;

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;
import org.apache.commons.math3.util.FastMath;

public class Agent {
    public Vector2D pos;
    public double speed; //position units per second
    private double orientation; //angle in radians, 0 is for only x axis movement
    public final double maxTurn = Math.toRadians(90); //maximum turning per second
    public final double maxSpeed = 50;

    Agent(double posX, double posY, double angle) {
        pos = new Vector2D(posX, posY);
        speed = 0;
        orientation = angle;
    }

    public double getOrientation() {
        return orientation;
    }

    public void turn(double angle) {
        orientation += angle;
        while (orientation > Math.PI) orientation -= 2*Math.PI;
        while (orientation < -Math.PI) orientation += 2*Math.PI;
    }

    public void moveForward(double moveFraction) {
        pos = pos.add(moveFraction * speed,
                new Vector2D(FastMath.cos(orientation), FastMath.sin(orientation))
        );
    }
}

File: src/main/scala/de/n2online/sonification/Agent.scala
package de.n2online.sonification

import de.n2online.sonification.Helpers._
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D

object Agent {
  val maxSpeed: Double = 20
  val maxTurnPerSecond: Double = Math.toRadians(60)
}

class Agent(posX: Double, posY: Double, initialOrientation: Double) {
  var pos: Vector2D = new Vector2D(posX, posY)
  var speed: Double = 0
  var targetDistance: Double = Double.NaN
  var targetAngle: Double = Double.NaN
  private var orientation: Double = initialOrientation

  def getOrientation: Double = orientation

  def turn(angle: Double) {
    orientation = wrapToSignedPi(orientation + angle)
  }

  def move(moveFraction: Double) {
    pos = pos.add(moveFraction * speed, new Vector2D(Math.cos(orientation), Math.sin(orientation)))
  }
}